pkg/etcd/migrate.go

```diff
@@ -0,0 +1,245 @@
+	cdcMetaVersion          = 1
+	etcdSessionTTL          = 10
+	noMetaVersion           = -1
+	migrateLogsWarnDuration = 5 * time.Second
+	migrationCampaignKey    = "ticdc-migration"
+)
+
+type keys map[string]string
+
+func (k keys) addPair(old, new string) {
+	k[old] = new
+}
+
+type migrator struct {
+	clusterID      string
+	oldMetaVersion int
+	newMetaVersion int
+	metaVersionKey string
+	// cdc old owner key
+	oldOwnerKey string
+	// etcd client
+	cli *Client
+	// all keyPrefixes needed to be migrated or update
+	// map from oldKeyPrefix to newKeyPrefix
+	keyPrefixes       keys
+	etcdNoMetaVersion bool
+}
+
+// Note: we do not use etcd transaction to migrate key
+// as it has the maximum operation limit in a single transaction.
+// So we use a double check mechanism to make sure the migration is complete.
+// 1. check and put metaVersion
+// 2. campaign old owner
+// 3. update keys
+// 4. update metaVersion
+func (m migrator) migrate(ctx context.Context) error {
+	// 1.1 check metaVersion, if the metaVersion in etcd does not match
+	// m.oldMetaVersion, it means that someone has migrated the meta data
+	metaVersion, err := getMetaVersion(ctx, m.cli, m.clusterID)
+	if err != nil {
+		log.Error("get meta version failed, etcd meta data migration failed", zap.Error(err))
+		return errors.Trace(err)
+	}
+
+	if metaVersion > m.newMetaVersion {
+		log.Panic("meta version in etcd is greater than the meta version in TiCDC",
+			zap.Int("etcdMetaVersion", metaVersion), zap.Int("cdcMetaVersion", m.newMetaVersion))
+	}
+
+	// if metaVersion in etcd is equal to m.newMetaVersion,
+	// it means that no need to migrate
+	if !m.etcdNoMetaVersion && metaVersion == m.newMetaVersion {
+		log.Warn("meta version no match, no need to migrate")
+		return nil
+	}
+
+	// 1.2 put metaVersionKey to etcd to panic old version cdc server
+	if m.etcdNoMetaVersion {
+		_, err := m.cli.Put(ctx, m.metaVersionKey, fmt.Sprintf("%d", m.oldMetaVersion))
+		if err != nil {
+			log.Error("put meta version failed, etcd meta data migration failed", zap.Error(err))
+			return errors.Trace(err)
+		}
+	}
+
+	// 3.campaign old owner to make sure old keys will not be updates
+	if err := m.campaignOldOwner(ctx); err != nil {
+		log.Error("campaign old owner failed, etcd meta data migration failed", zap.Error(err))
+		return errors.Trace(err)
+	}
+
+	// 4.campaign owner successfully, begin to migrate data
+	for oldPrefix, newPrefix := range m.keyPrefixes {
+		resp, err := m.cli.Get(ctx, oldPrefix, clientV3.WithPrefix())
+		if err != nil {
+			log.Error("get old meta data failed, etcd meta data migration failed", zap.Error(err))
+			return errors.Trace(err)
+		}
+		for _, v := range resp.Kvs {
+			oldKey := string(v.Key)
+			newKey := newPrefix + oldKey[len(oldPrefix):]
+			log.Info("migrate key", zap.String("oldKey", oldKey), zap.String("newKey", newKey))
+			// TODO(dongmen): we should update value after changefeedInfo struct altered
+			_, err := m.cli.Put(ctx, newKey, string(v.Value))
```

Comment:
Should we check lease atomically with the `Put` operation? Can the current node's lease suddenly expire doing these operations? If so, is data corruption possible?

Reply:
Normally, migrate will exit before cdc node's lease expired and only a cdc node with double ownership can put new meta data to etcd, so I consider there is no risk of data corruption.